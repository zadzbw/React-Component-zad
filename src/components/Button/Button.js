import React, {PropTypes, Component}from 'react';
import classNames from 'classnames';
import oneOf from '../../utils/oneOf';

import './Button.less';

const _type = {
	type: ['primary', 'ghost', 'dashed'],
	size: ['small', 'large'],
	shape: ['circle', 'circle-o']
};
const btnPrefix = 'zad-btn';

export default class Button extends Component {
	constructor(props) {
		super(props);
	}

	static propTypes = {
		type: PropTypes.string,
		size: PropTypes.string,
		shape: PropTypes.string,
		onClick: PropTypes.func
	};

	static defaultProps = {
		onClick: () => undefined
	};


	render() {
		const {type, size, shape, className, children, ...props} = this.props;

		const sizeClass = ({
				large: 'lg',
				small: 'sm',
			})[size] || '';

		const btnClass = classNames(btnPrefix, className, {
			[`${btnPrefix}-${type}`]: type && oneOf(_type.type, type),
			[`${btnPrefix}-${sizeClass}`]: size && oneOf(_type.size, size),
			[`${btnPrefix}-${shape}`]: shape && oneOf(_type.shape, shape)
		});

		return (
			<button
				{...props}
				type="button"
				className={btnClass}
			>
				<span>{children}</span>
			</button>
		);
	}
}
