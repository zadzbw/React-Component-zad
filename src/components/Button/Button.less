@import "../../style/variables";
@import "./ButtonMixin";

.zad-btn {
  .btn();
  .btn-default();

  > .fa + span, > span + .fa, > .fa + .fa {
    margin-left: 0.5em;
  }

  &-primary {
    .btn-primary();
  }

  &-ghost {
    .btn-ghost();
  }

  &-dashed {
    .btn-dashed();
  }

  &-circle {
    .btn-circle();
  }

  &-clicked:before {
    content: '';
    position: absolute;
    border-radius: inherit;
    border: 0 solid @primary-color;
    animation: buttonClick @animation-duration-slow ease-out forwards;
  }
}

@keyframes buttonClick {
  from {
    opacity: 1;
    top: -1px;
    left: -1px;
    bottom: -1px;
    right: -1px;
    border-width: 0;
  }

  to {
    opacity: 0;
    top: -6px;
    left: -6px;
    bottom: -6px;
    right: -6px;
    border-width: 6px;
  }
}