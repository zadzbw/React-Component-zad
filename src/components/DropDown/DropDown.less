@import "../../style/variables";

.zad-dropdown {
  display: inline-block;

  &-items {
    position: absolute;
    display: block;
    outline: none;
    margin-top: 8px;
    padding: 0;
    text-align: left;
    background-color: @body-background;
    border-radius: @border-radius-base;
    box-shadow: @box-shadow-base;
    z-index: @z-index-dropdown;
  }

  &-item {
    padding: @btn-padding-lg;
    margin: 0;
    font-size: @font-size-base;
    color: @text-color;
    white-space: nowrap;
    cursor: pointer;
    transition: all 0.3s ease;

    &:hover {
      background-color: tint(@primary-color, 85%);
    }
  }
}
