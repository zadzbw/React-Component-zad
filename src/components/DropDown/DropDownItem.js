/**
 * Created by zad on 16/12/20.
 */
import React, {PropTypes, Component} from 'react';

export default class DropDownItem extends Component {
  constructor(props) {
    super(props);
  }

  static propTypes = {
    name: PropTypes.string,
    onClick: PropTypes.func
  };

  static defaultProps = {
    name: '',
    onClick: () => undefined
  };

  render() {
    const {name, onClick} = this.props;
    return (
      <div onClick={onClick} className="zad-dropdown-item">
        {name}
      </div>
    );
  }
}