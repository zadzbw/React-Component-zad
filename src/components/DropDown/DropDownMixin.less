@import "../../style/variables";

.drop-down-items-wrap() {
  position: absolute;
  display: block;
  outline: none;
  margin: 0;
  padding: 0;
  text-align: left;
  background-color: @body-background;
  border-radius: @border-radius-base;
  box-shadow: @box-shadow-base;
  z-index: @z-index-dropdown;
}

.drop-down-item() {
  padding: @btn-padding-lg;
  margin: 0;
  font-size: @font-size-base;
  color: @text-color;
  white-space: nowrap;
  cursor: pointer;
  transition: all 0.3s ease;

  &:hover {
    background-color: tint(@primary-color, 85%);
  }
}
