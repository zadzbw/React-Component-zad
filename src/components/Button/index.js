/**
 * Created by zad on 16/11/30.
 */
import Button from './Button';

export default Button;