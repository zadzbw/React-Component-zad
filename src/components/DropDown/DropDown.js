/**
 * Created by zad on 16/12/20.
 */
import React, {PropTypes, Component} from 'react';
import Animate from 'rc-animate';
import classNames from 'classnames';
import chainFunc from '../../utils/chainFunc';
import Icon from '../Icon';
import Button from '../Button';
import ButtonGroup from '../ButtonGroup';
import DropDownItem from './DropDownItem';

import './DropDown.less';

export default class DropDown extends Component {
  static propTypes = {
    onSelect: PropTypes.func,
    menus: PropTypes.array
  };

  static defaultProps = {
    onSelect: () => undefined,
    menus: []
  };

  constructor(props) {
    super(props);
    this.state = {
      visible: false
    };
    this.switchVisible = this.switchVisible.bind(this);
    this.click = this.click.bind(this);
  }

  switchVisible() {
    this.setState({
      visible: !this.state.visible,
    }, () => {
      if (this.state.visible) {
        setTimeout(() => {
          window.addEventListener('click', this.click, false);
        }, 0);
      } else {
        setTimeout(() => {
          window.removeEventListener('click', this.click, false);
        }, 0);
      }
    });
  }

  click() {
    const {visible} = this.state;
    if (visible) {
      this.switchVisible();
    }
  }

  render() {
    const {className, style, children, menus, onSelect, ...props} = this.props;
    const dropDownClass = classNames('zad-dropdown', className);
    const Func = chainFunc(onSelect, this.switchVisible);
    const items = menus.map((menu) => {
      return (
        <DropDownItem onClick={Func.bind(this, menu.value)} name={menu.name} key={menu.value}/>
      );
    });
    const itemsWrap = !this.state.visible ? null : (
      <div className="zad-dropdown-items">
        {items}
      </div>
    );

    return (
      <div {...props} className={dropDownClass} style={style}>
        <ButtonGroup>
          <Button type={'ghost'}>{children}</Button>
          <Button type={'ghost'} onClick={this.switchVisible}><Icon name={'angle-down'}/></Button>
        </ButtonGroup>
        <Animate transitionName="slide-up">
          {itemsWrap}
        </Animate>
      </div>
    );
  }
}
