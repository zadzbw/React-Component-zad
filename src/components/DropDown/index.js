/**
 * Created by zad on 16/12/20.
 */
import DropDown from './DropDown';

export default DropDown;
