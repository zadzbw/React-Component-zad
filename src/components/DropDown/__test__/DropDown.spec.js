/**
 * Created by zad on 16/12/22.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import DropDown from '../../DropDown';
import DropDownItem from '../../DropDown/DropDownItem';
import Button from '../../Button';

const dropdown = [
  {name: '中国1-China', value: 'CN1'},
  {name: '日本1-Japan', value: 'JP1'},
  {name: '美国1-USA', value: 'USA1'},
  {
    name: '二级菜单', menus: [
    {name: '中国2-China', value: 'CN2'},
    {name: '日本2-Japan', value: 'JP2'},
    {name: '美国2-USA', value: 'USA2'},
    {
      name: '三级菜单', menus: [
      {name: '中国3-China', value: 'CN3'},
      {name: '日本3-Japan', value: 'JP3'},
      {name: '美国3-USA', value: 'USA3'}
    ]
    }
  ]
  }
];

describe('dropDown Test', () => {
  it('normal', () => {
    function test(value) {
      selected = value;
    }

    let selected = '';
    const wrapper = shallow(
      <DropDown menus={dropdown} onSelect={test}>
        DropDown
      </DropDown>
    );

    // open
    wrapper.find(Button).at(1).simulate('click');
    expect(wrapper.find(DropDownItem)).toHaveLength(11);
    // sub click
    wrapper.find(DropDownItem).at(3).simulate('click');
    wrapper.find(DropDownItem).at(7).simulate('click');
    expect(wrapper.find(DropDownItem)).toHaveLength(11);
    expect(selected).toEqual('');
    expect(wrapper.find(DropDownItem).at(3).shallow().find('.zad-dropdown-sub-title').text()).toEqual('二级菜单');
    expect(wrapper.find(DropDownItem).at(7).shallow().find('.zad-dropdown-sub-title').text()).toEqual('三级菜单');
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    // click
    wrapper.find(DropDownItem).at(0).simulate('click');
    expect(wrapper.find(DropDownItem)).toHaveLength(0);
    expect(selected).toEqual('CN1');

    // open
    wrapper.find(Button).at(1).simulate('click');
    // click
    wrapper.find(DropDownItem).at(5).simulate('click');
    expect(wrapper.find(DropDownItem)).toHaveLength(0);
    expect(selected).toEqual('JP2');

    // open
    wrapper.find(Button).at(1).simulate('click');
    // click
    wrapper.find(DropDownItem).at(10).simulate('click');
    expect(wrapper.find(DropDownItem)).toHaveLength(0);
    expect(selected).toEqual('USA3');
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
