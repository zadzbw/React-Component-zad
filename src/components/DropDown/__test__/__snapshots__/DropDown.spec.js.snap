exports[`dropDown Test normal 1`] = `
<div
  className="zad-dropdown">
  <ButtonGroup>
    <Button
      disabled={false}
      onClick={[Function]}
      type="ghost">
      DropDown
    </Button>
    <Button
      disabled={false}
      onClick={[Function]}
      type="ghost">
      <Icon
        name="angle-down" />
    </Button>
  </ButtonGroup>
  <Animate
    animation={Object {}}
    component="span"
    componentProps={Object {}}
    onAppear={[Function]}
    onEnd={[Function]}
    onEnter={[Function]}
    onLeave={[Function]}
    transitionAppear={false}
    transitionEnter={true}
    transitionLeave={true}
    transitionName="slide-up">
    <div
      className="zad-dropdown-items">
      <DropDownItem
        name="中国1-China"
        onClick={[Function]}
        type="normal" />
      <DropDownItem
        name="日本1-Japan"
        onClick={[Function]}
        type="normal" />
      <DropDownItem
        name="美国1-USA"
        onClick={[Function]}
        type="normal" />
      <DropDownItem
        name="二级菜单"
        onClick={[Function]}
        type="sub">
        <DropDownItem
          name="中国2-China"
          onClick={[Function]}
          type="normal" />
        <DropDownItem
          name="日本2-Japan"
          onClick={[Function]}
          type="normal" />
        <DropDownItem
          name="美国2-USA"
          onClick={[Function]}
          type="normal" />
        <DropDownItem
          name="三级菜单"
          onClick={[Function]}
          type="sub">
          <DropDownItem
            name="中国3-China"
            onClick={[Function]}
            type="normal" />
          <DropDownItem
            name="日本3-Japan"
            onClick={[Function]}
            type="normal" />
          <DropDownItem
            name="美国3-USA"
            onClick={[Function]}
            type="normal" />
        </DropDownItem>
      </DropDownItem>
    </div>
  </Animate>
</div>
`;

exports[`dropDown Test normal 2`] = `
<div
  className="zad-dropdown">
  <ButtonGroup>
    <Button
      disabled={false}
      onClick={[Function]}
      type="ghost">
      DropDown
    </Button>
    <Button
      disabled={false}
      onClick={[Function]}
      type="ghost">
      <Icon
        name="angle-down" />
    </Button>
  </ButtonGroup>
  <Animate
    animation={Object {}}
    component="span"
    componentProps={Object {}}
    onAppear={[Function]}
    onEnd={[Function]}
    onEnter={[Function]}
    onLeave={[Function]}
    transitionAppear={false}
    transitionEnter={true}
    transitionLeave={true}
    transitionName="slide-up" />
</div>
`;
